scripts/ci/testing/ci_run_airflow_testing.sh

```diff
@@ -170,5 +161,282 @@ do
+}
+
+# Checks status for all test types running. Returns 0 if any of the tests failed (with the exception of
+# Quarantined test.
+#
+# TEST_TYPES - all test types to run
+#    test_pids - map of test types to PIDs
+#
+# output:
+#   successful_tests - array of test types that succeeded
+#   failed_tests - array of test types that failed
+function check_test_types_status() {
+    start_end::group_start "Checking status for parallel tests: ${TEST_TYPES}"
+    local test_type_exit_code="0"
+    for test_type in "${!test_pids[@]}"; do
+        local pid="${test_pids[${test_type}]}"
+        wait "${pid}"
+        local temp_exit_code=$?
+        if [[ ${temp_exit_code} == "0" ]]; then
+            successful_tests+=("${test_type}")
+        else
+            failed_tests+=("${test_type}")
+            if [[ ${test_type} != "Quarantined" ]]; then
+                test_type_exit_code=${temp_exit_code}``
+            fi
+        fi
+    done
+    start_end::group_end
+    return "${test_type_exit_code}"
+}
+
+# Outputs logs for failed test type
+# $1 test type
+function output_log_for_failed_test_type(){
+    local failed_test_type=$1
+    local log_file="${CACHE_TMP_FILE_DIR}/output-${failed_test_type}.log"
+    start_end::group_start "${COLOR_RED}Output for failed ${failed_test_type} from log: ${log_file}${COLOR_RESET}"
+    echo "${COLOR_RED}##### Test type ${failed_test_type} failed ##### ${COLOR_RESET}"
+    echo
+    cat "${log_file}"
+    echo
+    echo "${COLOR_RED}##### Test type ${failed_test_type} failed ##### ${COLOR_RESET}"
+    echo
+    start_end::group_end
+}
+
+function print_test_summary() {
+    if [[ ${#successful_tests[@]} != 0 ]]; then
+        echo
+        echo "${COLOR_GREEN}Successful test types: ${successful_tests[*]} ${COLOR_RESET}"
+        echo
+    fi
+    if [[ ${#failed_tests[@]} != 0 ]]; then
+        echo
+        echo "${COLOR_RED}Failed test types: ${failed_tests[*]}${COLOR_RESET}"
+        echo
+        for test_type in "${failed_tests[@]}"
+        do
+            output_log_for_failed_test_type "${test_type}"
+        done
+    fi
+}
+
+# Runs all test types in parallel, monitors their progress and summarizes the result when finished.
+function run_all_test_types_in_parallel() {
+    local successful_tests=()
+    local failed_tests=()
+    local exit_code="0"
+    local log_file
+    local test_type
+    local test_pids
+    declare -A test_pids
+
+    start_end::group_start "Cleanup docker before parallel tests: ${TEST_TYPES}."
+    cleanup_docker
     start_end::group_end
-done
+
+    start_test_types_in_parallel "${@}"
+
+    set +e
+    monitor_test_types
+
+    check_test_types_status
+    exit_code=$?
+    set -e
+    print_test_summary
+    return ${exit_code}
+}
+
+function run_all_test_types_sequentially() {
+    local exit_code="0"
+    local successful_tests=()
+    local failed_tests=()
+    local log_file
+    local test_type
+    local pid
+    local test_pids
+    declare -A test_pids
```

Comment:
```suggestion
    local -A test_pids
```
(or `declare -A test_pids`) We don't need both:

> When used in a function, declare and typeset make each name local, as with the local command, unless the -g option is supplied.